proposals/downloads_get_file_hash.md

```diff
@@ -0,0 +1,161 @@
+# Proposal: `downloads.getFileHash()`
+
+**Summary**
+
+Allow extensions to query hashes of contents of downloaded files
+programmatically.
+
+**Document Metadata**
+
+**Author:** @bershanskyi
+
+**Sponsoring Browser:** Chromium
+
+**Contributors:** @Rob--W
+
+**Created:** 2024-04-08
+
+**Related Issues:** [WECG GitHub issue 401](https://github.com/w3c/webextensions/issues/401)
+
+## Motivation
+
+### Objective
+
+This API will enable extensions to privately and efficiently scan user
+downloads for "Known Unsafe" contents and validate the integrity of "Known Safe"
+files without exposing the raw content of the files.
+
+#### Use Cases
+
+##### Filtering for unwanted downloads
+
+Many antivirus products detect previously seen "safe" and "unsafe" files by
+computing a file content hash and checking it against a set of known hashes.
+These checks require minimal effort on the client side, yet are powerful
+enough to limit the spread of known malware downloads since effectively every
+client gets "inoculated" against a particular download as soon as a single
+antivirus detects a malicious file.
+User-Agents may store file hashes even after the original file is deleted to
+facilitate tracing of malware exposure even if the file is deemed to be
+dangerous after the user initially downloads it.
+
+##### Download integrity assurance
+
+Files can get damaged while they are shared on a file hosting server or NAS.
+While some file formats explicitly support integrity checks like embedded
+checksums or even cryptographic signatures, other files do not. This API
+would allow file managers or extensions to programmatically check the
+integrity of downloaded files if the file source elects to provide them.
+
+### Known Consumers
+
+None at the moment.
+
+## Specification
+
+### Schema
+
+
+``` idl
+namespace downloads {
+    callback GetFileHashCallback = void(DOMString hash);
+
+    [supportsPromises, permissions="downloads.getFileHash"]
+    static void getFileHash(
+        long downloadId,
+        [legalValues=("SHA-256")] DOMString algorithm,
+        GetFileHashCallback callback);
+}
+```
+
+``` idl
+namespace downloads {
+  enum FILE_HASH_ALGO {
+    "SHA-256"
```

Comment:
Not blocking, but I wonder how this is exposed as the constant, since there is no obvious approach as seen at https://github.com/w3c/webextensions/issues/563

- lowercase or upper case?
- what happens with the `-`?

Reply:
This is a good point and this proposal lacks clarity about enums. Here we actually have two hings to consider: the enum key and the enum value literal. This document currently does not specify the enum key explicitly. I copied enum value literal `"SHA-256"` from [Web Cryptography API §30.2 Registration](https://www.w3.org/TR/WebCryptoAPI/#sha-registration).

What do you think about this?
```idl
namespace downloads {
    enum FileHashAlgorithm {
        SHA256 = "SHA-256"
    }
}
```